File: sol/3439/3439.cpp
struct Solution {
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size(), prev  = 0;
        vector<int> gaps(n + 1);
        for (int i = 0; i < n; ++i){
            gaps[i] = startTime[i] - prev;
            prev = endTime[i];
        }
        gaps[n] = eventTime - prev;
        int w = 0;
        for (int i = 0; i <= k; ++i){
            w += gaps[i];
        }
        int res = w;
        for (int i = k + 1; i <= n; ++i){
            w += gaps[i] - gaps[i - k - 1];
            res = max(res, w);
        }
        return res;
    }
};

File: sol/3439/3439.py
class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: List[int], endTime: List[int]) -> int:
        intervals = []
        prev = 0
        for a, b in zip(startTime, endTime):
            intervals.append(a - prev)
            prev = b
        intervals.append(eventTime - prev)
        m = len(intervals)
        winSum = sum(intervals[:k + 1])
        res = winSum
        for i in range(k + 1, m):
            winSum += intervals[i] - intervals[i - k - 1]
            res = max(res, winSum)
        return res
